File: baml_language/sdk_tests/crates/cpp/function_calls/customizable/tests/test_generic_calls.cpp
// Generic function-call coverage (ns_generic_tests), explicit-binding form.
// Port of function_calls/customizable/test_generic_calls.py. Deviations:
// - Python's `fn[T](...)` subscript becomes explicit C++ template args
//   (`fn<int64_t>(...)`); both surfaces send the same named type-arg wire.
// - Generic statics need a parameterized class to be named in C++
//   (`GenericBox<int64_t>::new_(...)`), so the generated binding also sends
//   the class-level TypeVar (`T`) alongside the static's own (`V`); Python
//   sends only `V`. Same for NamedStatic<A,B,C>::make (A/B/C ride along).
//   `new` is a C++ keyword, so the emitter sanitizes it to `new_`.
// - Negative runtime cases are compile errors in C++ and are skipped:
//   test_generic_free_fn_requires_binding (a bare call to a template with no
//   deducible params does not compile), test_subscript_wrong_arity_raises
//   (template arity is checked at compile time), and
//   test_instance_method_unparameterized_receiver_raises (an unparameterized
//   GenericBox cannot be constructed - the template requires a type arg).
// - The reified `_type_args(obj)` assertions have no runtime counterpart:
//   the concrete type args ARE the function's static C++ return type.
#include <baml_sdk.hpp>
#include <baml_test.hpp>

namespace generic_tests = baml_sdk::generic_tests;
using generic_tests::ContainerShapes;
using generic_tests::GenericBox;
using generic_tests::GenericPair;
using generic_tests::GenericRecursive;
using generic_tests::GenericTriple;
using generic_tests::NamedStatic;
using generic_tests::StringIntPair;

// ===========================================================================
// basic cases, free functions
// ===========================================================================

BAML_TEST(identity_explicit) {
    BAML_ASSERT_EQ(generic_tests::identity<int64_t>(5), int64_t{5});
    BAML_ASSERT_EQ(generic_tests::identity<std::string>("hi"), std::string("hi"));

    const StringIntPair pair{"a", 1};
    BAML_ASSERT(generic_tests::identity<StringIntPair>(pair) == pair);

    const GenericBox<GenericBox<std::string>> box{GenericBox<std::string>{"hello"}};
    BAML_ASSERT((generic_tests::identity<GenericBox<GenericBox<std::string>>>(box)) == box);

    const GenericTriple<GenericBox<std::string>, double, bool> triple{
        GenericBox<std::string>{"hello"},
        {1.1, 2.2},
        {{"lorem", true}, {"ipsum", false}},
    };
    BAML_ASSERT(
        (generic_tests::identity<GenericTriple<GenericBox<std::string>, double, bool>>(triple)) ==
        triple);
}

BAML_TEST(identity_async_explicit) {
    BAML_ASSERT_EQ(generic_tests::identity_async<int64_t>(7).get(), int64_t{7});
}

BAML_TEST(tag_or_value_explicit) {
    // `tag_or_value` reflects its bound `T` back as a string; `x` must inhabit
    // the substituted `T | string | null`.
    BAML_ASSERT_EQ(generic_tests::tag_or_value<int64_t>(std::variant<int64_t, std::string>{
                       int64_t{5}}),
                   std::string("int"));
    // T=string makes the parameter variant<string, string>; the duplicate
    // alternative needs in_place_index (the engine dedups the union).
    BAML_ASSERT_EQ(generic_tests::tag_or_value<std::string>(std::variant<std::string, std::string>{
                       std::in_place_index<0>, "plain"}),
                   std::string("string"));
    const StringIntPair pair{"b", 2};
    const std::string tagged =
        generic_tests::tag_or_value<StringIntPair>(std::variant<StringIntPair, std::string>{pair});
    BAML_ASSERT(tagged.find("StringIntPair") != std::string::npos);
}

BAML_TEST(make_triple_explicit) {
    // A=int, B=string, C=bool, bound positionally by the template args.
    const GenericTriple<int64_t, std::string, bool> t =
        generic_tests::make_triple<int64_t, std::string, bool>(1, {"a", "b"}, {{"k", true}});
    BAML_ASSERT_EQ(t.first, int64_t{1});
    BAML_ASSERT(t.second == (std::vector<std::string>{"a", "b"}));
    BAML_ASSERT(t.third == (std::map<std::string, bool>{{"k", true}}));
}

// one_type_arg<T>() / two_type_args<A,B>(): return-position-only TypeVars.
// No argument carries `T`; the binding can only come from the template args.

BAML_TEST(one_type_arg_explicit) {
    BAML_ASSERT_EQ(generic_tests::one_type_arg<int64_t>(), std::string("int"));
    BAML_ASSERT_EQ(generic_tests::one_type_arg<std::string>(), std::string("string"));
    // Nested generic binding must encode fully (base class + concrete arg).
    const std::string nested = generic_tests::one_type_arg<GenericBox<int64_t>>();
    BAML_ASSERT(nested.find("GenericBox") != std::string::npos);
    BAML_ASSERT(nested.find("int") != std::string::npos);
}

BAML_TEST(two_type_args_explicit) {
    BAML_ASSERT_EQ((generic_tests::two_type_args<int64_t, std::string>()),
                   std::string("int | string"));
}

// test_generic_free_fn_requires_binding: skipped - a bare `one_type_arg()` /
// `two_type_args()` does not compile in C++ (no deducible template params).
// test_subscript_wrong_arity_raises: skipped - `two_type_args<int64_t>()` is a
// compile-time arity error in C++.

// ===========================================================================
// basic cases, generic classes
// ===========================================================================

BAML_TEST(consume_int_wrapper_baseline) {
    // No binding of any kind: a concretely-instantiated GenericBox<int64_t>
    // flows in and the int field flows back out.
    BAML_ASSERT_EQ(generic_tests::consume_int_wrapper(GenericBox<int64_t>{9}), int64_t{9});
}

BAML_TEST(genericbox_get_explicit) {
    // GenericBox<int64_t> carries the type arg; the binding encodes it as the
    // method frame's class-level T.
    const GenericBox<int64_t> b{5};
    BAML_ASSERT_EQ(b.get(), std::string("int"));
}

BAML_TEST(genericbox_pair_with_explicit) {
    // T from the GenericBox<int64_t> receiver, U from the method template arg.
    const GenericBox<int64_t> b{5};
    BAML_ASSERT_EQ(b.pair_with<std::string>("hello world"), std::string("int | string"));
}

BAML_TEST(genericbox_new_static_explicit) {
    const GenericBox<int64_t> box = GenericBox<int64_t>::new_<int64_t>(5);
    BAML_ASSERT_EQ(box.value, int64_t{5});
}

BAML_TEST(generic_static_infers_binding) {
    // The static's own V appears in a parameter (value: V), so C++ deduces it
    // with no explicit method template arg (deduction still sends the binding
    // explicitly on the wire - see the file header).
    const GenericBox<int64_t> box = GenericBox<int64_t>::new_(int64_t{5});
    BAML_ASSERT_EQ(box.value, int64_t{5});
}

BAML_TEST(named_static_distinct_typevar_names) {
    // Static TypeVar names (D, E) differ from the class's (A, B, C). C++ must
    // parameterize the class to name the static, so arbitrary A/B/C bindings
    // ride along (Python sends none); the named wire slots D/E by name.
    BAML_ASSERT_EQ((NamedStatic<int64_t, int64_t, int64_t>::make<int64_t, std::string>(1, "x")),
                   std::string("int | string"));
}

// test_instance_method_unparameterized_receiver_raises: skipped - C++ cannot
// construct an unparameterized GenericBox (the class template requires a type
// argument), so the negative case is a compile error.

BAML_TEST(extract_explicit) {
    // Nested generic: extract<A,B,C,D>(a: GenericPair<GenericPair<A,B>,
    // GenericPair<C,D>>).
    const GenericPair<GenericPair<int64_t, std::string>, GenericPair<bool, double>> pair{
        GenericPair<int64_t, std::string>{1, "a"},
        GenericPair<bool, double>{true, 1.5},
    };
    BAML_ASSERT_EQ((generic_tests::extract<int64_t, std::string, bool, double>(pair)),
                   std::string("int | string | bool | float"));
}

// ===========================================================================
// basic case, T in return position only
// ===========================================================================

BAML_TEST(parse_as_explicit) {
    // T bound by the host via the template arg.
    const StringIntPair pair =
        generic_tests::parse_as<StringIntPair>("{\"my_string\": \"x\", \"my_int\": 3}");
    BAML_ASSERT(pair == (StringIntPair{"x", 3}));
    BAML_ASSERT_EQ(generic_tests::parse_as<int64_t>("42"), int64_t{42});
}

// ===========================================================================
// complex cases
// ===========================================================================

BAML_TEST(second_of_explicit) {
    BAML_ASSERT_EQ(generic_tests::second_of<std::string>(
                       GenericPair<int64_t, std::string>{1, "hi"}),
                   std::string("hi"));
    const StringIntPair pair{"z", 9};
    const GenericPair<int64_t, StringIntPair> p{0, pair};
    BAML_ASSERT(generic_tests::second_of<StringIntPair>(p) == pair);
}

BAML_TEST(list_head_explicit) {
    const GenericRecursive<int64_t> linked_list{
        7, GenericRecursive<int64_t>{8, std::nullopt}};
    BAML_ASSERT_EQ(generic_tests::list_head<int64_t>(linked_list), int64_t{7});
}

BAML_TEST(choose_explicit) {
    BAML_ASSERT_EQ(generic_tests::choose<int64_t>(1, 2), int64_t{1});
    BAML_ASSERT_EQ(generic_tests::choose<std::string>("a", "b"), std::string("a"));
}

BAML_TEST(read_items_explicit) {
    const ContainerShapes<int64_t> container{
        1,
        {1, 2, 3},
        {{"k", 4}},
        std::nullopt,
        std::nullopt,
    };
    BAML_ASSERT(generic_tests::read_items<int64_t>(container) == (std::vector<int64_t>{1, 2, 3}));
}

// ===========================================================================
// outbound generics
// ===========================================================================

BAML_TEST(wrap_explicit) {
    const GenericBox<int64_t> w = generic_tests::wrap<int64_t>(5);
    BAML_ASSERT_EQ(w.value, int64_t{5});
}

// ===========================================================================
// reified generics returned by NON-generic functions
// ===========================================================================
// The outbound mirror of consume_int_wrapper: the callee's return type pins
// the class type args at the definition site. In C++ the reified type args
// are the static return type itself, so Python's __pydantic_generic_metadata__
// assertions are the (compile-checked) declarations below.

BAML_TEST(make_int_box_reified) {
    const GenericBox<int64_t> box = generic_tests::make_int_box();
    BAML_ASSERT_EQ(box.value, int64_t{7});
}

BAML_TEST(make_int_container_reified) {
    // The single int binding is reified into every field shape (bare, list,
    // map, optional, union).
    const ContainerShapes<int64_t> c = generic_tests::make_int_container();
    BAML_ASSERT_EQ(c.item, int64_t{1});
    BAML_ASSERT(c.items == (std::vector<int64_t>{1, 2, 3}));
    BAML_ASSERT(c.by_key == (std::map<std::string, int64_t>{{"k", 4}}));
    BAML_ASSERT(!c.maybe.has_value());
    BAML_ASSERT(c.mixed == (std::optional<std::variant<int64_t, std::string>>{
                    std::variant<int64_t, std::string>{int64_t{5}}}));
}

BAML_TEST(make_nested_box_reified) {
    // The type arg is itself a generic instance; the inner GenericBox decodes
    // out of the outer one's field.
    const GenericBox<GenericBox<int64_t>> outer = generic_tests::make_nested_box();
    BAML_ASSERT_EQ(outer.value.value, int64_t{9});
}

BAML_TEST(make_int_str_bool_triple_reified) {
    // Multiple TypeVars reified across mixed field shapes.
    const GenericTriple<int64_t, std::string, bool> t =
        generic_tests::make_int_str_bool_triple();
    BAML_ASSERT_EQ(t.first, int64_t{1});
    BAML_ASSERT(t.second == (std::vector<std::string>{"a", "b"}));
    BAML_ASSERT(t.third == (std::map<std::string, bool>{{"k", true}}));
}

File: baml_language/sdk_tests/crates/java/function_calls/customizable/TestGenericCalls.java
// Generic *function-call* coverage (ns_generic_tests) — the EXPLICIT-binding
// variant.
//
// Port of python_pydantic2/function_calls/customizable/test_generic_calls.py.
//
// ===========================================================================
// INVENTED API SHAPE (flagged prominently in the report — Java has no `_types=`
// kwarg or `fn[T]` subscript, and the state-of-completeness doc leaves the
// explicit-type-args surface "TBD"). This port proposes ONE minimal, consistent
// shape and uses it everywhere:
//
//   * `baml_bridge.BamlType` — a type token:
//       BamlType.INT / STRING / BOOL / FLOAT      (primitive tokens)
//       BamlType.of(StringIntPair.class)          (user class / enum)
//       BamlType.of(GenericBox.class, BamlType.INT) (reified generic)
//   * `baml_bridge.BamlTypes` — a NAMED type-arg binding set (the wire form the
//     engine actually uses; also covers partial binding, replacing Python's
//     `_types=`):
//       BamlTypes.of("T", BamlType.INT)
//       BamlTypes.of("A", BamlType.INT).and("B", BamlType.STRING)
//   * Explicit generic CALLS take a trailing `BamlTypes` argument on a dedicated
//     overload: `Fns.identity(5L, BamlTypes.of("T", BamlType.INT))`,
//     `b.pair_with("x", BamlTypes.of("U", BamlType.STRING))`,
//     `GenericBox.new$(5L, BamlTypes.of("V", BamlType.INT))`.
//   * Generic VALUE classes:
//       - unbound instance (no reified type args): plain ctor `new GenericBox<>(5L)`;
//       - bound/reified instance: static factory taking the reified type
//         token(s) positionally, then the fields:
//         `GenericBox.of(BamlType.INT, 5L)`,
//         `GenericPair.of(BamlType.INT, BamlType.STRING, 1L, "hi")`;
//       - `instance.bamlTypeArgs()` -> `List<BamlType>` reads the reified type
//         args back (the analog of Python's
//         `__pydantic_generic_metadata__["args"]`).
//   * BAML reserved-word method `new` escapes to `new$` (the doc's `$` escaping
//     precedent; also flagged).
//   * Engine type-mismatch rejections surface as native `IllegalArgumentException`
//     (state doc: `baml.errors.TypeMismatch` -> `IllegalArgumentException`,
//     mirroring Python's `TypeError`).
// ===========================================================================

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import baml_bridge.BamlType;
import baml_bridge.BamlTypes;
import baml_sdk.generic_tests.ContainerShapes;
import baml_sdk.generic_tests.Fns;
import baml_sdk.generic_tests.GenericBox;
import baml_sdk.generic_tests.GenericPair;
import baml_sdk.generic_tests.GenericRecursive;
import baml_sdk.generic_tests.GenericTriple;
import baml_sdk.generic_tests.NamedStatic;
import baml_sdk.generic_tests.StringIntPair;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class TestGenericCalls {

    // =======================================================================
    // basic cases, free functions
    // =======================================================================

    @Test
    void test_generic_calls_identity_explicit() {
        assertEquals(5L, Fns.identity(5L, BamlTypes.of("T", BamlType.INT)));
        assertEquals("hi", Fns.identity("hi", BamlTypes.of("T", BamlType.STRING)));

        StringIntPair pair = new StringIntPair("a", 1L);
        assertEquals(pair, Fns.identity(pair, BamlTypes.of("T", BamlType.of(StringIntPair.class))));

        GenericBox<GenericBox<String>> box =
                GenericBox.of(
                        BamlType.of(GenericBox.class, BamlType.STRING),
                        GenericBox.of(BamlType.STRING, "hello"));
        assertEquals(
                box,
                Fns.identity(
                        box,
                        BamlTypes.of(
                                "T",
                                BamlType.of(
                                        GenericBox.class,
                                        BamlType.of(GenericBox.class, BamlType.STRING)))));

        GenericTriple<GenericBox<String>, Double, Boolean> triple =
                GenericTriple.of(
                        BamlType.of(GenericBox.class, BamlType.STRING),
                        BamlType.FLOAT,
                        BamlType.BOOL,
                        GenericBox.of(BamlType.STRING, "hello"),
                        List.of(1.1, 2.2),
                        Map.of("lorem", true, "ipsum", false));
        assertEquals(
                triple,
                Fns.identity(
                        triple,
                        BamlTypes.of(
                                "T",
                                BamlType.of(
                                        GenericTriple.class,
                                        BamlType.of(GenericBox.class, BamlType.STRING),
                                        BamlType.FLOAT,
                                        BamlType.BOOL))));
    }

    @Test
    void test_generic_calls_identity_async_explicit() {
        assertEquals(7L, Fns.identity_async(7L, BamlTypes.of("T", BamlType.INT)).join());
    }

    @Test
    void test_generic_calls_tag_or_value_explicit() {
        // `tag_or_value` reflects its bound `T` back as a string; `x` must
        // inhabit the substituted `T | string | null` (a union arg -> `Object`).
        assertEquals("int", Fns.tag_or_value(5L, BamlTypes.of("T", BamlType.INT)));
        assertEquals("string", Fns.tag_or_value("plain", BamlTypes.of("T", BamlType.STRING)));
        StringIntPair pair = new StringIntPair("b", 2L);
        assertTrue(
                Fns.tag_or_value(pair, BamlTypes.of("T", BamlType.of(StringIntPair.class)))
                        .contains("StringIntPair"));
    }

    @Test
    void test_generic_calls_make_triple_explicit() {
        // A=int, B=str, C=bool, bound positionally by name.
        var t =
                Fns.make_triple(
                        1L,
                        List.of("a", "b"),
                        Map.of("k", true),
                        BamlTypes.of("A", BamlType.INT)
                                .and("B", BamlType.STRING)
                                .and("C", BamlType.BOOL));
        assertInstanceOf(GenericTriple.class, t);
        assertEquals(1L, t.first());
        assertEquals(List.of("a", "b"), t.second());
        assertEquals(Map.of("k", true), t.third());
    }

    @Test
    void test_generic_calls_one_type_arg_explicit() {
        assertEquals("int", Fns.one_type_arg(BamlTypes.of("T", BamlType.INT)));
        assertEquals("string", Fns.one_type_arg(BamlTypes.of("T", BamlType.STRING)));
        // Nested generic binding must encode fully (base class + concrete arg).
        String nested =
                Fns.one_type_arg(BamlTypes.of("T", BamlType.of(GenericBox.class, BamlType.INT)));
        assertTrue(nested.contains("GenericBox") && nested.contains("int"));
    }

    @Test
    void test_generic_calls_two_type_args_explicit() {
        assertEquals(
                "int | string",
                Fns.two_type_args(BamlTypes.of("A", BamlType.INT).and("B", BamlType.STRING)));
    }

    @Test
    void test_generic_calls_generic_free_fn_requires_binding() {
        // `one_type_arg<T>()` / `two_type_args<A,B>()` are return/body-only: NO
        // argument carries the TypeVar, so a bare call finds no evidence and the
        // engine (Gate A) rejects it. `baml.errors.TypeMismatch` surfaces as a
        // native `IllegalArgumentException` (Python's `TypeError`).
        IllegalArgumentException excOne =
                assertThrows(IllegalArgumentException.class, () -> Fns.one_type_arg());
        assertTrue(excOne.getMessage().contains("could not infer a type"));
        assertTrue(excOne.getMessage().contains("one_type_arg"));
        IllegalArgumentException excTwo =
                assertThrows(IllegalArgumentException.class, () -> Fns.two_type_args());
        assertTrue(excTwo.getMessage().contains("could not infer a type"));
    }

    @Test
    void test_generic_calls_subscript_wrong_arity_raises() {
        // java-port note: Python's subscript form requires FULL arity and raises
        // a host-side message "expected 2 type argument(s) for ['A', 'B'], got 1".
        // Java has a single `BamlTypes` surface that (like Python's `_types=`)
        // permits partial binding, so there is no separate subscript arity check.
        // Binding only "A" for `two_type_args<A,B>` leaves body-only `B` with no
        // evidence, so the engine rejects it (could-not-infer) — the meaningful
        // Java analog of the under-specified call.
        assertThrows(
                IllegalArgumentException.class,
                () -> Fns.two_type_args(BamlTypes.of("A", BamlType.INT)));
    }

    // =======================================================================
    // basic cases, generic classes
    // =======================================================================

    @Test
    void test_generic_calls_consume_int_wrapper_baseline() {
        // A concretely-instantiated `GenericBox<int>` flows in; the `int` field
        // flows back out. Anchors the suite.
        assertEquals(9L, Fns.consume_int_wrapper(GenericBox.of(BamlType.INT, 9L)));
    }

    @Test
    void test_generic_calls_genericbox_get_explicit() {
        // `GenericBox.of(BamlType.INT, ...)` carries the type arg; the host
        // recovers it from the receiver as the method frame's class-level `T`.
        GenericBox<Long> b = GenericBox.of(BamlType.INT, 5L);
        assertEquals("int", b.get());
    }

    @Test
    void test_generic_calls_genericbox_pair_with_explicit() {
        // `T` from the reified receiver, `U` from the method's explicit binding.
        GenericBox<Long> b = GenericBox.of(BamlType.INT, 5L);
        assertEquals("int | string", b.pair_with("hello world", BamlTypes.of("U", BamlType.STRING)));
    }

    @Test
    void test_generic_calls_genericbox_new_static_explicit() {
        // Generic STATIC method (`new` -> `new$`); only the static's own `V` is
        // bound, no class type args ride along.
        GenericBox<Long> box = GenericBox.new$(5L, BamlTypes.of("V", BamlType.INT));
        assertInstanceOf(GenericBox.class, box);
        assertEquals(5L, box.value());
    }

    @Test
    void test_generic_calls_generic_static_infers_binding() {
        // A generic static's own `V` appears in a parameter, so a bare call
        // INFERS it from the value — no explicit binding needed.
        GenericBox<Long> box = GenericBox.new$(5L);
        assertInstanceOf(GenericBox.class, box);
        assertEquals(5L, box.value());
    }

    @Test
    void test_generic_calls_named_static_distinct_typevar_names() {
        // Static TypeVar names differ from the class's (`make<D,E>`); each binds
        // by NAME into the static frame's own params.
        assertEquals(
                "int | string",
                NamedStatic.make(1L, "x", BamlTypes.of("D", BamlType.INT).and("E", BamlType.STRING)));
    }

    @Test
    void test_generic_calls_instance_method_unparameterized_receiver_raises() {
        // java-port note: `new GenericBox<>(5L)` is UNBOUND (no reified type
        // args), so `pair_with`'s class `T` can't be recovered host-side. Python
        // raises a pydantic-specific message; the Java analog is a host-side
        // `IllegalArgumentException` about the receiver needing to be reified.
        GenericBox<Long> unbound = new GenericBox<>(5L);
        IllegalArgumentException exc =
                assertThrows(
                        IllegalArgumentException.class,
                        () -> unbound.pair_with("x", BamlTypes.of("U", BamlType.STRING)));
        assertTrue(exc.getMessage().contains("receiver"));
    }

    @Test
    void test_generic_calls_extract_explicit() {
        // Nested generic; body is `reflect.type_of<A|B|C|D>()`.
        GenericPair<GenericPair<Long, String>, GenericPair<Boolean, Double>> pair =
                GenericPair.of(
                        BamlType.of(GenericPair.class, BamlType.INT, BamlType.STRING),
                        BamlType.of(GenericPair.class, BamlType.BOOL, BamlType.FLOAT),
                        GenericPair.of(BamlType.INT, BamlType.STRING, 1L, "a"),
                        GenericPair.of(BamlType.BOOL, BamlType.FLOAT, true, 1.5));
        assertEquals(
                "int | string | bool | float",
                Fns.extract(
                        pair,
                        BamlTypes.of("A", BamlType.INT)
                                .and("B", BamlType.STRING)
                                .and("C", BamlType.BOOL)
                                .and("D", BamlType.FLOAT)));
    }

    // =======================================================================
    // basic case, `T` in return position only
    // =======================================================================

    @Test
    void test_generic_calls_parse_as_explicit() {
        // `T` bound by the host explicitly. Return-position-only, so the call
        // site needs a target/witness type for the erased `<T>`.
        StringIntPair pair =
                Fns.parse_as(
                        "{\"my_string\": \"x\", \"my_int\": 3}",
                        BamlTypes.of("T", BamlType.of(StringIntPair.class)));
        assertEquals(new StringIntPair("x", 3L), pair);
        assertEquals(42L, Fns.<Long>parse_as("42", BamlTypes.of("T", BamlType.INT)));
    }

    // =======================================================================
    // complex cases
    // =======================================================================

    @Test
    void test_generic_calls_second_of_explicit() {
        assertEquals(
                "hi",
                Fns.second_of(
                        GenericPair.of(BamlType.INT, BamlType.STRING, 1L, "hi"),
                        BamlTypes.of("T", BamlType.STRING)));
        StringIntPair pair = new StringIntPair("z", 9L);
        GenericPair<Long, StringIntPair> p =
                GenericPair.of(BamlType.INT, BamlType.of(StringIntPair.class), 0L, pair);
        assertEquals(pair, Fns.second_of(p, BamlTypes.of("T", BamlType.of(StringIntPair.class))));
    }

    @Test
    void test_generic_calls_list_head_explicit() {
        GenericRecursive<Long> linkedList =
                GenericRecursive.of(
                        BamlType.INT, 7L, GenericRecursive.of(BamlType.INT, 8L, null));
        assertEquals(7L, Fns.list_head(linkedList, BamlTypes.of("T", BamlType.INT)));
    }

    @Test
    void test_generic_calls_choose_explicit() {
        assertEquals(1L, Fns.choose(1L, 2L, BamlTypes.of("T", BamlType.INT)));
        assertEquals("a", Fns.choose("a", "b", BamlTypes.of("T", BamlType.STRING)));
    }

    @Test
    void test_generic_calls_read_items_explicit() {
        ContainerShapes<Long> container =
                ContainerShapes.of(BamlType.INT, 1L, List.of(1L, 2L, 3L), Map.of("k", 4L), null, null);
        assertEquals(List.of(1L, 2L, 3L), Fns.read_items(container, BamlTypes.of("T", BamlType.INT)));
    }

    // =======================================================================
    // outbound generics
    // =======================================================================

    @Test
    void test_generic_calls_wrap_explicit() {
        GenericBox<Long> w = Fns.wrap(5L, BamlTypes.of("T", BamlType.INT));
        assertInstanceOf(GenericBox.class, w);
        assertEquals(5L, w.value());
    }

    // =======================================================================
    // reified generics returned by NON-generic functions
    // =======================================================================

    @Test
    void test_generic_calls_make_int_box_reified() {
        GenericBox<Long> box = Fns.make_int_box();
        assertInstanceOf(GenericBox.class, box);
        assertEquals(List.of(BamlType.INT), box.bamlTypeArgs());
        assertEquals(7L, box.value());
    }

    @Test
    void test_generic_calls_make_int_container_reified() {
        ContainerShapes<Long> c = Fns.make_int_container();
        assertInstanceOf(ContainerShapes.class, c);
        assertEquals(List.of(BamlType.INT), c.bamlTypeArgs());
        assertEquals(1L, c.item());
        assertEquals(List.of(1L, 2L, 3L), c.items());
        assertEquals(Map.of("k", 4L), c.by_key());
        org.junit.jupiter.api.Assertions.assertNull(c.maybe());
        assertEquals(5L, c.mixed());
    }

    @Test
    void test_generic_calls_make_nested_box_reified() {
        GenericBox<GenericBox<Long>> outer = Fns.make_nested_box();
        assertInstanceOf(GenericBox.class, outer);
        // The outer box's single type arg is itself the reified `GenericBox<int>`.
        assertEquals(List.of(BamlType.of(GenericBox.class, BamlType.INT)), outer.bamlTypeArgs());
        GenericBox<Long> inner = outer.value();
        assertInstanceOf(GenericBox.class, inner);
        assertEquals(List.of(BamlType.INT), inner.bamlTypeArgs());
        assertEquals(9L, inner.value());
    }

    @Test
    void test_generic_calls_make_int_str_bool_triple_reified() {
        GenericTriple<Long, String, Boolean> t = Fns.make_int_str_bool_triple();
        assertInstanceOf(GenericTriple.class, t);
        assertEquals(List.of(BamlType.INT, BamlType.STRING, BamlType.BOOL), t.bamlTypeArgs());
        assertEquals(1L, t.first());
        assertEquals(List.of("a", "b"), t.second());
        assertEquals(Map.of("k", true), t.third());
    }
}
